Enhancements needed to implement Solaris-specific features:
 16734033 sparc: X server enhancement for multi-X session and failover
 17039591 sparc M series: X auto-configuration for multi-session operation

not suitable for upstream

--- xorg-server-21.1.20/config/hal.c
+++ xorg-server-21.1.20/config/hal.c
@@ -1,6 +1,7 @@
 /*
  * Copyright © 2007 Daniel Stone
  * Copyright © 2007 Red Hat, Inc.
+ * Copyright (c) 2013, 2025, Oracle and/or its affiliates.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
@@ -40,6 +41,22 @@ #include "hotplug.h"
 #include "config-backends.h"
 #include "os.h"
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined(SUNSOFT))
+#include <sys/stat.h>
+#include <unistd.h>
+
+#define MAX_DEVICES    4
+
+DeviceIntPtr added_devices[MAX_DEVICES];
+int num_added_devices = 0;
+Bool abort_on_fail_over = FALSE;
+static Bool do_abort = FALSE;
+
+extern int  num_total_disp_dev;
+extern int  num_session_disp_dev;
+extern char disp_dev_path[PATH_MAX];
+#endif
+
 #define LIBHAL_PROP_KEY "input.x11_options."
 #define LIBHAL_XKB_PROP_KEY "input.xkb."
 
@@ -124,6 +141,51 @@ }
     return ret;
 }
 
+#if defined(SUNSOFT) && (defined(__sparc__) || defined(__sparc))
+static Bool check_inactive_session(char const *path) {
+    struct stat statbuf;
+    char linkpath[PATH_MAX];
+    ssize_t readstatus;
+    char *usbpath = NULL;
+    char *ptr;
+    char disppath[PATH_MAX];
+
+    if ((num_session_disp_dev == num_total_disp_dev) || !disp_dev_path[0])
+       return FALSE;
+
+    if (lstat(path, &statbuf) == 0 &&
+               (statbuf.st_mode & S_IFMT) == S_IFLNK) {
+       readstatus = readlink(path, linkpath, sizeof(linkpath));
+
+       if (readstatus > 0 && readstatus < sizeof(linkpath)) {
+           linkpath[readstatus] = 0;
+           usbpath = linkpath;
+           if (strncmp(usbpath, "../..", sizeof("../..") - 1) == 0)
+               usbpath += sizeof("../..") - 1;
+           if (strncmp(usbpath, "/devices", sizeof("/devices") - 1) == 0)
+               usbpath += sizeof("/devices") - 1;
+       }
+    }
+
+    if (!usbpath)
+       return FALSE;
+
+    if ((ptr = strchr(usbpath + 1, '/')) != NULL)
+       *ptr = 0;
+    else
+       return FALSE;
+
+    strlcpy(disppath, disp_dev_path, sizeof(disppath));
+
+    if ((ptr = strchr(disppath + 1, '/')) != NULL)
+       *ptr = 0;
+    else
+       return FALSE;
+
+    return (strcmp(usbpath, disppath));
+}
+#endif
+
 static void
 device_added(LibHalContext * hal_ctx, const char *udi)
 {
@@ -136,6 +198,11 @@ DBusError error;
     struct xkb_options xkb_opts = { 0 };
     int rc;
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined(SUNSOFT))
+    if (do_abort)
+       return;
+#endif
+
     LibHalPropertySet *set = NULL;
     LibHalPropertySetIterator set_iter;
     char *psi_key = NULL, *tmp_val;
@@ -234,6 +301,28 @@ input_options = input_option_new(input_o
     input_options = input_option_new(input_options, "driver", driver);
     input_options = input_option_new(input_options, "name", name);
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined(SUNSOFT))
+    if (!strcmp(name, "keyboard") || !strcmp(name, "mouse")) {
+       if (check_inactive_session(path)) {
+           if (abort_on_fail_over) {
+               /* M5: Input devices were removed, new input device added is to
+                  activate another session, reset it.
+                */
+               do_abort = TRUE;
+               LogMessage(X_INFO, "config/hal: Server to abort\n");
+           } else
+               /* M5: No removal of input devices happened, new input device
+                  added is to activate another session, do nothing.
+                */
+               LogMessage(X_INFO, "config/hal: Not adding input device %s\n", name);
+
+            goto unwind;
+       } else
+           /* M5: new input device added is to activate current session. */
+           abort_on_fail_over = FALSE;
+    }
+#endif
+
     if (asprintf(&config_info, "hal:%s", udi) == -1) {
         config_info = NULL;
         LogMessage(X_ERROR, "config/hal: couldn't allocate name\n");
@@ -404,6 +493,26 @@ dev = NULL;
         goto unwind;
     }
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined(SUNSOFT))
+    if ((num_session_disp_dev < num_total_disp_dev) &&
+               (!strcmp(name, "keyboard") || !strcmp(name, "mouse"))) {
+       int i;
+
+       if (num_added_devices == MAX_DEVICES) {
+           LogMessage(X_ERROR, "config/hal: Too manay devices to add\n");
+           goto unwind;
+       }
+
+       for (i = 0; i < MAX_DEVICES; i++) {
+           if (added_devices[i] == 0) {
+               added_devices[i] = dev;
+               num_added_devices++;
+               break;
+           }
+       }
+    }
+#endif
+
  unwind:
     if (set)
         libhal_free_property_set(set);
@@ -436,6 +545,13 @@ free(xkb_opts.options);
 
     dbus_error_free(&error);
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined(SUNSOFT))
+    if (do_abort) {
+       config_fini();
+       GiveUp(0);
+    }
+#endif
+
     return;
 }
 
--- xorg-server-21.1.20/hw/xfree86/common/xf86Xinput.c
+++ xorg-server-21.1.20/hw/xfree86/common/xf86Xinput.c
@@ -45,6 +45,9 @@ * and author(s) shall not be used in adv
  * the sale, use or other dealings in this Software without prior written
  * authorization from the copyright holder(s) and author(s).
  */
++/*
++ * Copyright (c) 2013, Oracle and/or its affiliates. All Rights Reserved.
++ */
 
 #ifdef HAVE_XORG_CONFIG_H
 #include <xorg-config.h>
@@ -124,6 +127,15 @@ .next = &new_input_devices_list,
     .prev = &new_input_devices_list,
 };
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined(sun))
+#define MAX_DEVICES            4
+extern int  num_total_disp_dev;
+extern int  num_session_disp_dev;
+extern DeviceIntPtr added_devices[MAX_DEVICES];
+extern int num_added_devices;
+extern Bool abort_on_fail_over;
+#endif
+
 /**
  * Eval config and modify DeviceVelocityRec accordingly
  */
@@ -1553,6 +1565,25 @@ else {
         SendDevicePresenceEvent(dev->id, DeviceUnrecoverable);
         DeleteInputDeviceRequest(dev);
     }
+
+#if ((defined(__sparc__) || defined(__sparc)) && defined(sun))
+    if (num_session_disp_dev < num_total_disp_dev) {
+       int i;
+
+       for (i = 0; i < MAX_DEVICES; i++) {
+           if (added_devices[i] == dev) {
+               added_devices[i] = 0;
+               if (--num_added_devices == 0)
+                   /* M5: will abort server when another X session
+                      is activated.
+                    */
+                   abort_on_fail_over = TRUE;
+
+               break;
+           }
+       }
+    }
+#endif
 }
 
 /**
--- xorg-server-21.1.20/hw/xfree86/common/xf86pciBus.c
+++ xorg-server-21.1.20/hw/xfree86/common/xf86pciBus.c
@@ -54,6 +54,12 @@ #define PCI_VENDOR_GENERIC		0x00FF
 /* Bus-specific globals */
 int pciSlotClaimed = 0;
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+int  num_total_disp_dev = 0;
+int  num_session_disp_dev = 0;
+char disp_dev_path[PATH_MAX];
+#endif
+
 #define PCIINFOCLASSES(c) \
     ( (((c) & 0x00ff0000) == (PCI_CLASS_PREHISTORIC << 16)) \
       || (((c) & 0x00ff0000) == (PCI_CLASS_DISPLAY << 16)) \
@@ -115,6 +121,11 @@ primaryBus.type = BUS_PCI;
                 primaryBus.id.pci = info;
             }
             info->user_data = 0;
+
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+            if (IS_VGA(info->device_class))
+               num_total_disp_dev++;
+#endif
         }
     }
     free(iter);
@@ -488,6 +499,16 @@ const struct pci_id_match *const devices
     GDevPtr *devList;
     const unsigned numDevs = xf86MatchDevice(drvp->driverName, &devList);
 
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+    struct sol_device_private {
+       struct pci_device  base;
+       const char * device_string;
+    };
+#define DEV_PATH(dev)    (((struct sol_device_private *) dev)->device_string)
+
+    num_session_disp_dev = numDevs;
+#endif
+
     for (i = 0; i < numDevs; i++) {
         struct pci_device_iterator *iter;
         unsigned device_id;
@@ -564,6 +585,11 @@ if (entry != -1) {
                     if ((*drvp->PciProbe) (drvp, entry, pPci,
                                            devices[j].match_data)) {
                         foundScreen = TRUE;
+
+#if ((defined(__sparc__) || defined(__sparc)) && defined (sun))
+                        strlcpy(disp_dev_path, DEV_PATH(pPci), sizeof(disp_dev_path));
+#endif
+
                     }
                     else
                         xf86UnclaimPciSlot(pPci, devList[i]);
